src/index.ts

```diff
@@ -9,65 +10,145 @@ class VPCPlugin {
+        const f = service.functions[fName];
+        const fVPCDiscovery = f.vpcDiscovery;
+
+        if (typeof fVPCDiscovery === "boolean" && !fVPCDiscovery) {
+          // skip vpc setup for `vpcDiscovery=false` option
+          Globals.logInfo(`Skipping VPC config for the function '${fName}'`);
+          continue;
+        }
+
+        const vpcDiscovery = basicVPCDiscovery;
+        // check vpcDiscovery config
+        if (fVPCDiscovery) {
+          const noSubnetsAndGroups = fVPCDiscovery.subnetNames == null && fVPCDiscovery.securityGroupNames == null;
+          // validate vpcDiscovery config options
+          if (fVPCDiscovery.vpcName == null || noSubnetsAndGroups) {
+            Globals.logWarning(
+              `The function '${fName}' is not configured correctly.` +
+              "Please see README for proper setup. The basic vpc config are applied"
+            );
+          } else {
+            // merge basic config to specific for the function
+            Object.assign(vpcDiscovery, await this.getVpcConfig(fVPCDiscovery));
+          }
+        }
+
+        // init vpc empty config in case not exists
+        f.vpc = f.vpc || {};
+
+        // log warning in case vpc.subnetIds and vpcDiscovery.subnetNames are specified.
+        if (f.vpc.subnetIds && fVPCDiscovery.subnetNames) {
+          Globals.logWarning(
+            `vpc.subnetIds' are specified for the function '${fName}' 
+            and overrides 'vpcDiscovery.subnetNames' discovery config.`
+          );
+        }
 
-    const service = this.serverless.service
+        // set vpc.subnetIds
+        if (!f.vpc.subnetIds && vpcDiscovery.subnetIds) {
+          f.vpc.subnetIds = vpcDiscovery.subnetIds;
+        }
 
-    const vpcId = await this.ec2Wrapper.getVpcId(service.custom.vpc.vpcName)
-    const subnetIds = await this.ec2Wrapper.getSubnetIds(vpcId, service.custom.vpc.subnetNames)
-    const securityGroupIds = await this.ec2Wrapper.getSecurityGroupIds(vpcId, service.custom.vpc.securityGroupNames)
+        // log warning in case vpc.securityGroupIds and vpcDiscovery.securityGroupNames are specified.
+        if (f.vpc.securityGroupIds && fVPCDiscovery.securityGroupNames) {
+          Globals.logWarning(
+            `vpc.securityGroupIds' are specified for the function '${fName}' 
+            and overrides 'vpcDiscovery.securityGroupNames' discovery config.`
+          );
+        }
 
-    service.provider.vpc = {
-      subnetIds: subnetIds,
-      securityGroupIds: securityGroupIds
+        // set vpc.securityGroupIds
+        if (!f.vpc.securityGroupIds && vpcDiscovery.securityGroupIds) {
+          f.vpc.securityGroupIds = vpcDiscovery.securityGroupIds;
+        }
+      }
     }
 
-    return service.provider.vpc
+    return service.functions;
+  }
+
+  /**
+   * Gets the desired vpc with the designated subnets and security groups
+   * that were set in serverless config file
+   * @returns {Promise<object>}
```

Comment:
```suggestion
   * @returns {Promise<VPC>}
```

Reply:
Currently, we return `service.functions` object